### ISSUE TITLE
AuthZ Engine: Authorization Decision Stages

### ISSUE THREAD

**Issue by seantleonard** [Azure/data-api-builder#301]
Input depends on #300 

Given a request and permissions definitions, make a decision about whether the request (action/entity/columns) should be allowed or denied.

|Processing Stage| Description | HTTP Failure Code | HTTP SubStatus Code |
| --- | :--- | :---: | :---: |
|Check role context of request| The X-Hawaii-Role HTTP header matches a role in the User’s Role claims. | 403 Forbidden| 100 Role Membership mismatch |
|Check role is configured on entity| entity[entityName].roles.ContainsKey(roleName)| 403 Forbidden | 101 Entity not in scope of role|
|Check action allowed for role| entity[entityName].roles[roleName].Actions.ContainsKey(actionType) | 403 Forbidden| 102 Action on Entity not in scope of role |
|Check that requested columns are allowed for action| Columns in request must match entity:role:action enabled columns| 403 Forbidden  | 103 Column Mismatch |
